### ISSUE TITLE
Diffs to make compatible with linuxodbc & newer releases of Go

### ISSUE THREAD

**Issue by AllanCochrane** [weigj/go-odbc#1]
Basically run gofix on odbc.go, comment out the conversion to complex and change calls to SQLColAttribute

``` go
diff --git a/odbc.go b/odbc.go
index 0af70bf..d6bc16c 100644
--- a/odbc.go
+++ b/odbc.go
@@ -148,7 +148,7 @@ func (conn *Connection) Prepare(sql string, params ...interface{}) (*Statement,
 }

 //func (conn *Connection) Query(sql string, params ...interface{}) (stmt *Statement, err *ODBCError) {
-// //TODO 
+// //TODO
 //}

 func (conn *Connection) Commit() (err *ODBCError) {
@@ -334,12 +334,12 @@ type Row struct {
 // Get(Columnindex)
 // TODO Get(ColumnName)
 func (r *Row) Get(a interface{}) interface{} {
-   value := reflect.NewValue(a)
-   switch f := value.(type) {
-   case *reflect.IntValue:
-       return r.Data[f.Get()]
-   case *reflect.UintValue:
-       return r.Data[f.Get()]
+   value := reflect.ValueOf(a)
+   switch f := value; f.Kind() {
+   case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+       return r.Data[f.Int()]
+   case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64, reflect.Uintptr:
+       return r.Data[f.Uint()]
        //  case *reflect.StringValue:
        //      i := r.Meta[f.Get()]
        //      return r.Data[i]
@@ -349,42 +349,42 @@ func (r *Row) Get(a interface{}) interface{} {

 func (r *Row) GetInt(a interface{}) (ret int64) {
    v := r.Get(a)
-   value := reflect.NewValue(v)
-   switch f := value.(type) {
-   case *reflect.IntValue:
-       ret = int64(f.Get())
-   case *reflect.UintValue:
-       ret = int64(f.Get())
+   value := reflect.ValueOf(v)
+   switch f := value; f.Kind() {
+   case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+       ret = int64(f.Int())
+   case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64, reflect.Uintptr:
+       ret = int64(f.Uint())
    }
    return
 }

 func (r *Row) GetFloat(a interface{}) (ret float64) {
    v := r.Get(a)
-   value := reflect.NewValue(v)
-   switch f := value.(type) {
-   case *reflect.FloatValue:
-       ret = float64(f.Get())
+   value := reflect.ValueOf(v)
+   switch f := value; f.Kind() {
+   case reflect.Float32, reflect.Float64:
+       ret = float64(f.Float())
    }
    return
 }

-func (r *Row) GetComplex(a interface{}) (ret complex) {
-   v := r.Get(a)
-   value := reflect.NewValue(v)
-   switch f := value.(type) {
-   case *reflect.ComplexValue:
-       ret = complex(f.Get())
-   }
-   return
-}
+// func (r *Row) GetComplex(a interface{}) (ret complex) {
+//     v := r.Get(a)
+//     value := reflect.NewValue(v)
+//     switch f := value.(type) {
+//     case *reflect.ComplexValue:
+//         ret = complex(f.Get())
+//     }
+//     return
+// }

 func (r *Row) GetString(a interface{}) (ret string) {
    v := r.Get(a)
-   value := reflect.NewValue(v)
-   switch f := value.(type) {
-   case *reflect.StringValue:
-       ret = f.Get()
+   value := reflect.ValueOf(v)
+   switch f := value; f.Kind() {
+   case reflect.String:
+       ret = f.String()
    }
    return
 }
@@ -421,14 +421,14 @@ func (stmt *Statement) FetchOne() (*Row, *ODBCError) {
 }

 func (stmt *Statement) GetField(field_index int) (v interface{}, ftype int, flen int, err *ODBCError) {
-   var field_type C.int
-   var field_len C.int
+   var field_type C.SQLINTEGER
+   var field_len C.SQLINTEGER
    var ll C.SQLSMALLINT
-   ret := C.SQLColAttribute(C.SQLHSTMT(stmt.handle), C.SQLUSMALLINT(field_index+1), C.SQL_DESC_CONCISE_TYPE, C.SQLPOINTER(unsafe.Pointer(uintptr(0))), C.SQLSMALLINT(0), &ll, C.SQLPOINTER(unsafe.Pointer(&field_type)))
+   ret := C.SQLColAttribute(C.SQLHSTMT(stmt.handle), C.SQLUSMALLINT(field_index+1), C.SQL_DESC_CONCISE_TYPE, C.SQLPOINTER(unsafe.Pointer(uintptr(0))), C.SQLSMALLINT(0), &ll, &field_type)
    if !Success(ret) {
        debugPrint("GetFiled type Error")
    }
-   ret = C.SQLColAttribute(C.SQLHSTMT(stmt.handle), C.SQLUSMALLINT(field_index+1), C.SQL_DESC_LENGTH, C.SQLPOINTER(unsafe.Pointer(uintptr(0))), C.SQLSMALLINT(0), &ll, C.SQLPOINTER(unsafe.Pointer(&field_len)))
+   ret = C.SQLColAttribute(C.SQLHSTMT(stmt.handle), C.SQLUSMALLINT(field_index+1), C.SQL_DESC_LENGTH, C.SQLPOINTER(unsafe.Pointer(uintptr(0))), C.SQLSMALLINT(0), &ll, &field_len)
    if !Success(ret) {
        debugPrint("GetFiled len Error")
    }
@@ -500,7 +500,7 @@ func (stmt *Statement) BindParam(index int, param interface{}) *ODBCError {
    var ParameterValuePtr C.SQLPOINTER
    var BufferLength C.SQLLEN
    var StrLen_or_IndPt C.SQLLEN
-   v := reflect.NewValue(param)
+   v := reflect.ValueOf(param)
    if param == nil {
        ft, _, _, _, err := stmt.GetParamType(index)
        if err != nil {
@@ -514,12 +514,12 @@ func (stmt *Statement) BindParam(index int, param interface{}) *ODBCError {
        StrLen_or_IndPt = C.SQL_NULL_DATA
        ColumnSize = 1
    } else {
-       switch v := v.(type) {
-       case *reflect.BoolValue:
+       switch v.Kind() {
+       case reflect.Bool:
            ParameterType = C.SQL_BIT
            ValueType = C.SQL_C_BIT
            var b [1]byte
-           if v.Get() {
+           if v.Bool() {
                b[0] = 1
            } else {
                b[0] = 0
@@ -527,37 +527,37 @@ func (stmt *Statement) BindParam(index int, param interface{}) *ODBCError {
            ParameterValuePtr = C.SQLPOINTER(unsafe.Pointer(&b[0]))
            BufferLength = 1
            StrLen_or_IndPt = 0
-       case *reflect.IntValue:
+       case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
            switch v.Type().Kind() {
            case reflect.Int:
            case reflect.Int8, reflect.Int16, reflect.Int32:
                ParameterType = C.SQL_INTEGER
                ValueType = C.SQL_C_LONG
-               var l C.long = C.long(v.Get())
+               var l C.long = C.long(v.Int())
                ParameterValuePtr = C.SQLPOINTER(unsafe.Pointer(&l))
                BufferLength = 4
                StrLen_or_IndPt = 0
            case reflect.Int64:
                ParameterType = C.SQL_BIGINT
                ValueType = C.SQL_C_SBIGINT
-               var ll C.longlong = C.longlong(v.Get())
+               var ll C.longlong = C.longlong(v.Int())
                ParameterValuePtr = C.SQLPOINTER(unsafe.Pointer(&ll))
                BufferLength = 8
                StrLen_or_IndPt = 0
            }
-       case *reflect.FloatValue:
+       case reflect.Float32, reflect.Float64:
            ParameterType = C.SQL_DOUBLE
            ValueType = C.SQL_C_DOUBLE
-           var d C.double = C.double(v.Get())
+           var d C.double = C.double(v.Float())
            ParameterValuePtr = C.SQLPOINTER(unsafe.Pointer(&d))
            BufferLength = 8
            StrLen_or_IndPt = 0
-       case *reflect.ComplexValue:
-       case *reflect.StringValue:
-           var slen C.SQLUINTEGER = C.SQLUINTEGER(len(v.Get()))
+       case reflect.Complex64, reflect.Complex128:
+       case reflect.String:
+           var slen C.SQLUINTEGER = C.SQLUINTEGER(len(v.String()))
            ParameterType = C.SQL_VARCHAR
            ValueType = C.SQL_C_CHAR
-           s := []byte(v.Get())
+           s := []byte(v.String())
            ParameterValuePtr = C.SQLPOINTER(unsafe.Pointer(&s[0]))
            ColumnSize = slen
            BufferLength = C.SQLINTEGER(slen + 1)

```

**Comment by AllanCochrane** [comment]
This seems to work with unixodbc from Ubuntu: 2.2.14p2-2ubuntu1 installed in Ubuntu 11.04 talking to a Vertica database through Vertica-supplied drivers.

**Comment by bwheatley** [comment]
I'm having trouble trying to apply the patch. Thought i could just do a git apply -v (FILE with your above diff) but it spit that out.

**Comment by bwheatley** [comment]
blah even after gofix i'm still getting

```
 [/home/bill/go-odbc]# gomake install
CGOPKGPATH= cgo --  odbc.go 
touch _obj/_cgo_run
6g  -o _go_.6 util.go  _obj/odbc.cgo1.go _obj/_cgo_gotypes.go
odbc.go:243[_obj/odbc.cgo1.go:246]: cannot use &nor (type *_Ctypedef_SQLINTEGER) as type *_Ctypedef_SQLLEN in function argument
odbc.go:342[_obj/odbc.cgo1.go:345]: use of builtin complex not in function call
odbc.go:347[_obj/odbc.cgo1.go:350]: use of builtin complex not in function call
odbc.go:347[_obj/odbc.cgo1.go:350]: missing argument to complex - complex(f.Complex(), <nil>)
odbc.go:397[_obj/odbc.cgo1.go:400]: cannot use _Ctypedef_SQLPOINTER(unsafe.Pointer(&field_type)) (type _Ctypedef_SQLPOINTER) as type *_Ctypedef_SQLLEN in function argument
odbc.go:401[_obj/odbc.cgo1.go:404]: cannot use _Ctypedef_SQLPOINTER(unsafe.Pointer(&field_len)) (type _Ctypedef_SQLPOINTER) as type *_Ctypedef_SQLLEN in function argument
odbc.go:425[_obj/odbc.cgo1.go:428]: cannot use _Ctypedef_SQLINTEGER(int(field_len) + 4) (type _Ctypedef_SQLINTEGER) as type _Ctypedef_SQLLEN in function argument
odbc.go:430[_obj/odbc.cgo1.go:433]: cannot use _Ctypedef_SQLINTEGER(int(field_len) + 4) (type _Ctypedef_SQLINTEGER) as type _Ctypedef_SQLLEN in function argument
odbc.go:532[_obj/odbc.cgo1.go:535]: cannot use slen (type _Ctypedef_SQLUINTEGER) as type _Ctypedef_SQLULEN in assignment
odbc.go:533[_obj/odbc.cgo1.go:536]: cannot use _Ctypedef_SQLINTEGER(slen + 1) (type _Ctypedef_SQLINTEGER) as type _Ctypedef_SQLLEN in assignment
odbc.go:533[_obj/odbc.cgo1.go:536]: too many errors
make: *** [_go_.6] Error 1
```

**Comment by AllanCochrane** [comment]
Hmm.

So gofix sorts out the reflect changes and I just commented out the complex function (line 350 in your error message).

The only other fix I applied manually was this:
-   var field_type C.int
-   var field_len C.int
-   var field_type C.SQLINTEGER
-   var field_len C.SQLINTEGER
  var ll C.SQLSMALLINT
-   ret := C.SQLColAttribute(C.SQLHSTMT(stmt.handle), C.SQLUSMALLINT(field_index+1), C.SQL_DESC_CONCISE_TYPE, C.SQLPOINTER(unsafe.Pointer(uintptr(0))), C.SQLSMALLINT(0), &ll, C.SQLPOINTER(unsafe.Pointer(&field_type)))
-   ret := C.SQLColAttribute(C.SQLHSTMT(stmt.handle), C.SQLUSMALLINT(field_index+1), C.SQL_DESC_CONCISE_TYPE, C.SQLPOINTER(unsafe.Pointer(uintptr(0))), C.SQLSMALLINT(0), &ll, &field_type)
  if !Success(ret) {
      debugPrint("GetFiled type Error")
  }
-   ret = C.SQLColAttribute(C.SQLHSTMT(stmt.handle), C.SQLUSMALLINT(field_index+1), C.SQL_DESC_LENGTH, C.SQLPOINTER(unsafe.Pointer(uintptr(0))), C.SQLSMALLINT(0), &ll, C.SQLPOINTER(unsafe.Pointer(&field_len)))
-   ret = C.SQLColAttribute(C.SQLHSTMT(stmt.handle), C.SQLUSMALLINT(field_index+1), C.SQL_DESC_LENGTH, C.SQLPOINTER(unsafe.Pointer(uintptr(0))), C.SQLSMALLINT(0), &ll, &field_len)

Basically change the types of the variables field_type and field_len to C.SQLINTEGER and then replace the calls to C.SQLPOINTER(unsafe.Pointer(&field_len)) with &field_len (similar for field_type) which cause the errors around line 400 in your error message.

I'll see if I can get git to provide a patch compatible output.

**Comment by AllanCochrane** [comment]
The diff seems to apply cleanly via git apply for me. A standard diff is below. Let me know if you need more info and I can paste the whole file somewhere.

151c151

## < //    //TODO 

> //    //TODO
> 337,342c337,342
> <   value := reflect.NewValue(a)
> <   switch f := value.(type) {
> <   case *reflect.IntValue:
> <       return r.Data[f.Get()]
> <   case *reflect.UintValue:
> 
> ## <       return r.Data[f.Get()]
> 
>   value := reflect.ValueOf(a)
>   switch f := value; f.Kind() {
>   case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
>       return r.Data[f.Int()]
>   case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64, reflect.Uintptr:
>       return r.Data[f.Uint()]
> 352,357c352,357
> <   value := reflect.NewValue(v)
> <   switch f := value.(type) {
> <   case *reflect.IntValue:
> <       ret = int64(f.Get())
> <   case *reflect.UintValue:
> 
> ## <       ret = int64(f.Get())
> 
>   value := reflect.ValueOf(v)
>   switch f := value; f.Kind() {
>   case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
>       ret = int64(f.Int())
>   case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64, reflect.Uintptr:
>       ret = int64(f.Uint())
> 364,367c364,367
> <   value := reflect.NewValue(v)
> <   switch f := value.(type) {
> <   case *reflect.FloatValue:
> 
> ## <       ret = float64(f.Get())
> 
>   value := reflect.ValueOf(v)
>   switch f := value; f.Kind() {
>   case reflect.Float32, reflect.Float64:
>       ret = float64(f.Float())
> 372,380c372,380
> < func (r *Row) GetComplex(a interface{}) (ret complex) {
> <   v := r.Get(a)
> <   value := reflect.NewValue(v)
> <   switch f := value.(type) {
> <   case *reflect.ComplexValue:
> <       ret = complex(f.Get())
> <   }
> <   return
> 
> ## < }
> 
> // func (r *Row) GetComplex(a interface{}) (ret complex) {
> //    v := r.Get(a)
> //    value := reflect.NewValue(v)
> //    switch f := value.(type) {
> //    case *reflect.ComplexValue:
> //        ret = complex(f.Get())
> //    }
> //    return
> // }
> 384,387c384,387
> <   value := reflect.NewValue(v)
> <   switch f := value.(type) {
> <   case *reflect.StringValue:
> 
> ## <       ret = f.Get()
> 
>   value := reflect.ValueOf(v)
>   switch f := value; f.Kind() {
>   case reflect.String:
>       ret = f.String()
> 424,425c424,425
> <   var field_type C.int
> 
> ## <   var field_len C.int
> 
>   var field_type C.SQLINTEGER
>   var field_len C.SQLINTEGER
> 427c427
> 
> ## <   ret := C.SQLColAttribute(C.SQLHSTMT(stmt.handle), C.SQLUSMALLINT(field_index+1), C.SQL_DESC_CONCISE_TYPE, C.SQLPOINTER(unsafe.Pointer(uintptr(0))), C.SQLSMALLINT(0), &ll, C.SQLPOINTER(unsafe.Pointer(&field_type)))
> 
>   ret := C.SQLColAttribute(C.SQLHSTMT(stmt.handle), C.SQLUSMALLINT(field_index+1), C.SQL_DESC_CONCISE_TYPE, C.SQLPOINTER(unsafe.Pointer(uintptr(0))), C.SQLSMALLINT(0), &ll, &field_type)
> 431c431
> 
> ## <   ret = C.SQLColAttribute(C.SQLHSTMT(stmt.handle), C.SQLUSMALLINT(field_index+1), C.SQL_DESC_LENGTH, C.SQLPOINTER(unsafe.Pointer(uintptr(0))), C.SQLSMALLINT(0), &ll, C.SQLPOINTER(unsafe.Pointer(&field_len)))
> 
>   ret = C.SQLColAttribute(C.SQLHSTMT(stmt.handle), C.SQLUSMALLINT(field_index+1), C.SQL_DESC_LENGTH, C.SQLPOINTER(unsafe.Pointer(uintptr(0))), C.SQLSMALLINT(0), &ll, &field_len)
> 503c503
> 
> ## <   v := reflect.NewValue(param)
> 
>   v := reflect.ValueOf(param)
> 517,518c517,518
> <       switch v := v.(type) {
> 
> ## <       case *reflect.BoolValue:
> 
> ```
>   switch v.Kind() {
>   case reflect.Bool:
> ```
> 
> 522c522
> 
> ## <           if v.Get() {
> 
> ```
>       if v.Bool() {
> ```
> 
> 530c530
> 
> ## <       case *reflect.IntValue:
> 
> ```
>   case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
> ```
> 
> 536c536
> 
> ## <               var l C.long = C.long(v.Get())
> 
> ```
>           var l C.long = C.long(v.Int())
> ```
> 
> 543c543
> 
> ## <               var ll C.longlong = C.longlong(v.Get())
> 
> ```
>           var ll C.longlong = C.longlong(v.Int())
> ```
> 
> 548c548
> 
> ## <       case *reflect.FloatValue:
> 
> ```
>   case reflect.Float32, reflect.Float64:
> ```
> 
> 551c551
> 
> ## <           var d C.double = C.double(v.Get())
> 
> ```
>       var d C.double = C.double(v.Float())
> ```
> 
> 555,557c555,557
> <       case *reflect.ComplexValue:
> <       case *reflect.StringValue:
> 
> ## <           var slen C.SQLUINTEGER = C.SQLUINTEGER(len(v.Get()))
> 
> ```
>   case reflect.Complex64, reflect.Complex128:
>   case reflect.String:
>       var slen C.SQLUINTEGER = C.SQLUINTEGER(len(v.String()))
> ```
> 
> 560c560
> 
> ## <           s := []byte(v.Get())
> 
> ```
>       s := []byte(v.String())
> ```

**Comment by AllanCochrane** [comment]
See     git://gist.github.com/996425.git

**Comment by bwheatley** [comment]
cool thank you i'll give it a try now. :)

**Comment by bwheatley** [comment]
So even after manually patching the file since it wouldn't take for whatever reason i'm still getting

```
 [/home/bill/go-odbc]# make
6g  -o _go_.6 util.go  _obj/odbc.cgo1.go _obj/_cgo_gotypes.go
odbc.go:243[_obj/odbc.cgo1.go:246]: cannot use &nor (type *_Ctypedef_SQLINTEGER) as type *_Ctypedef_SQLLEN in function argument
odbc.go:388[_obj/odbc.cgo1.go:391]: cannot use &field_type (type *_Ctypedef_SQLINTEGER) as type *_Ctypedef_SQLLEN in function argument
odbc.go:392[_obj/odbc.cgo1.go:395]: cannot use &field_len (type *_Ctypedef_SQLINTEGER) as type *_Ctypedef_SQLLEN in function argument
odbc.go:416[_obj/odbc.cgo1.go:419]: cannot use _Ctypedef_SQLINTEGER(int(field_len) + 4) (type _Ctypedef_SQLINTEGER) as type _Ctypedef_SQLLEN in function argument
odbc.go:421[_obj/odbc.cgo1.go:424]: cannot use _Ctypedef_SQLINTEGER(int(field_len) + 4) (type _Ctypedef_SQLINTEGER) as type _Ctypedef_SQLLEN in function argument
odbc.go:523[_obj/odbc.cgo1.go:526]: cannot use slen (type _Ctypedef_SQLUINTEGER) as type _Ctypedef_SQLULEN in assignment
odbc.go:524[_obj/odbc.cgo1.go:527]: cannot use _Ctypedef_SQLINTEGER(slen + 1) (type _Ctypedef_SQLINTEGER) as type _Ctypedef_SQLLEN in assignment
odbc.go:525[_obj/odbc.cgo1.go:528]: cannot use _Ctypedef_SQLINTEGER(slen) (type _Ctypedef_SQLINTEGER) as type _Ctypedef_SQLLEN in assignment
make: *** [_go_.6] Error 1
```

**Comment by bwheatley** [comment]
I have the libs too i think

```
bill@ns1 [/home/bill/go-odbc]# rpm -qa | grep -ri odbc
unixODBC-2.2.11-7.1
unixODBC-devel-2.2.11-7.1
bill@ns1 [/home/bill/go-odbc]# rpm -qa | grep -ri freetds
freetds-devel-0.64-11.el5.centos
freetds-0.64-11.el5.centos
```

**Comment by bwheatley** [comment]
OH i did also fix the issue with the complex block they changed it to require complex32/complex64

```
func (stmt *Statement) BindCol(col int, data interface{}, buff_indicator int) *ODBCError {
    // TODO
    switch f := data.(type) {
    case bool: //1 SQL_BIT
    case int: //4
    case string:
    case float: //4
    case float64: //8
    case complex64: //8
    default:
    }
    //ret := C.SQLBindCol(C.SQLHSTMT(stmt.handle), col, C.SQL_C_CHAR, &data, len(data), &buff_indicator)
    return nil
}
```

So that block works now. Though it's unrelated to my previous error.

**Comment by bwheatley** [comment]
Even tried it on my ubuntu natty box to see if i'd have similar luck. But same errors very weird.

**Comment by AllanCochrane** [comment]
Hi

Will post whole file as gist later

Allan

**Comment by bwheatley** [comment]
cool thanks :)

**Comment by AllanCochrane** [comment]
Hopefully this helps

git://gist.github.com/998355.git

https://gist.github.com/998355

https://gist.github.com/998355Allan

**Comment by bwheatley** [comment]
thanks man it's exactly the same as mine.

```
bill@bill-ubuntu:~/git/go-odbc$ make
6g  -o _go_.6 util.go  _obj/odbc.cgo1.go _obj/_cgo_gotypes.go
odbc.go:243[_obj/odbc.cgo1.go:246]: cannot use &nor (type *_Ctypedef_SQLINTEGER) as type *_Ctypedef_SQLLEN in function argument
odbc.go:388[_obj/odbc.cgo1.go:391]: cannot use &field_type (type *_Ctypedef_SQLINTEGER) as type *_Ctypedef_SQLLEN in function argument
odbc.go:392[_obj/odbc.cgo1.go:395]: cannot use &field_len (type *_Ctypedef_SQLINTEGER) as type *_Ctypedef_SQLLEN in function argument
odbc.go:416[_obj/odbc.cgo1.go:419]: cannot use _Ctypedef_SQLINTEGER(int(field_len) + 4) (type _Ctypedef_SQLINTEGER) as type _Ctypedef_SQLLEN in function argument
odbc.go:421[_obj/odbc.cgo1.go:424]: cannot use _Ctypedef_SQLINTEGER(int(field_len) + 4) (type _Ctypedef_SQLINTEGER) as type _Ctypedef_SQLLEN in function argument
odbc.go:523[_obj/odbc.cgo1.go:526]: cannot use slen (type _Ctypedef_SQLUINTEGER) as type _Ctypedef_SQLULEN in assignment
odbc.go:524[_obj/odbc.cgo1.go:527]: cannot use _Ctypedef_SQLINTEGER(slen + 1) (type _Ctypedef_SQLINTEGER) as type _Ctypedef_SQLLEN in assignment
odbc.go:525[_obj/odbc.cgo1.go:528]: cannot use _Ctypedef_SQLINTEGER(slen) (type _Ctypedef_SQLINTEGER) as type _Ctypedef_SQLLEN in assignment
make: *** [_go_.6] Error 1
```

wonder if i'm missing any of the unixodbc packages that you have?

```
bill@bill-ubuntu:~/git/go-odbc$ dpkg -l | grep -riE "unixodbc|freetds"
ii  freetds-bin                                                                   0.82-7                                     FreeTDS command-line utilities
ii  freetds-common                                                                0.82-7                                     configuration files for FreeTDS SQL client libraries
ii  freetds-dev                                                                   0.82-7                                     MS SQL and Sybase client library (static libs and headers)
ii  unixodbc                                                                      2.2.14p2-2ubuntu1                          ODBC tools libraries
ii  unixodbc-bin                                                                  2.2.14p2-2ubuntu1                          Graphical tools for ODBC management and browsing
ii  unixodbc-dev                                                                  2.2.14p2-2ubuntu1                          ODBC libraries for UNIX (development files)
```

Btw thanks appreciate the help figuring it out.

**Comment by AllanCochrane** [comment]
dpkg -l | grep -riE "unixodbc|freetds"
ii  unixodbc                              2.2.14p2-2ubuntu1
         ODBC tools libraries
ii  unixodbc-dev                          2.2.14p2-2ubuntu1
         ODBC libraries for UNIX (development files)

Hope that helps

Allan

**Comment by bwheatley** [comment]
hrm what version of go do you have? because we're exactly the same so maybe that is an issue too. :)

**Comment by AllanCochrane** [comment]
8g -V

8g version release.r57.1 8294

 gcc -v

Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/i386-linux-gnu/gcc/i686-linux-gnu/4.5.2/lto-wrapper
Target: i686-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Ubuntu/Linaro
4.5.2-8ubuntu4' --with-bugurl=file:///usr/share/doc/gcc-4.5/README.Bugs
--enable-languages=c,c++,fortran,objc,obj-c++ --prefix=/usr
--program-suffix=-4.5 --enable-shared --enable-multiarch
--with-multiarch-defaults=i386-linux-gnu --enable-linker-build-id
--with-system-zlib --libexecdir=/usr/lib/i386-linux-gnu
--without-included-gettext --enable-threads=posix
--with-gxx-include-dir=/usr/include/c++/4.5 --libdir=/usr/lib/i386-linux-gnu
--enable-nls --with-sysroot=/ --enable-clocale=gnu --enable-libstdcxx-debug
--enable-libstdcxx-time=yes --enable-plugin --enable-gold
--enable-ld=default --with-plugin-ld=ld.gold --enable-objc-gc
--enable-targets=all --disable-werror --with-arch-32=i686
--with-tune=generic --enable-checking=release --build=i686-linux-gnu
--host=i686-linux-gnu --target=i686-linux-gnu
Thread model: posix
gcc version 4.5.2 (Ubuntu/Linaro 4.5.2-8ubuntu4)
